Feature/parse given step on feature level

ggladko97 opened the pull request:
[WIP]
Load background for Gherkin Feature.

YevheniiHladkevych committed:
Added parsing GIVEN on feature scope

Signed-off-by: YevheniiHladkevych <yhladkevych@networkedassets.org>

YevheniiHladkevych committed:
Add access to ast.gherkin Features in Loader

Signed-off-by: YevheniiHladkevych <yhladkevych@networkedassets.org>

YevheniiHladkevych committed:
Add Jvm Static annotation

Signed-off-by: YevheniiHladkevych <yhladkevych@networkedassets.org>

YevheniiHladkevych committed:
Load background steps

- parse background data to the 2D array

Signed-off-by: YevheniiHladkevych <yhladkevych@networkedassets.org>

YevheniiHladkevych committed:
Load background steps

- add Background to GherkinFeature data schema
- put Background to the FeatureSpecification

Signed-off-by: YevheniiHladkevych <yhladkevych@networkedassets.org>

YevheniiHladkevych committed:
Load background steps

- remove unused method
- change return datatype to [][] array
- "background" is placed on "bindings" level

Signed-off-by: YevheniiHladkevych <yhladkevych@networkedassets.org>

YevheniiHladkevych committed:
Fixed loading background steps

- update Reflections

Signed-off-by: YevheniiHladkevych <yhladkevych@networkedassets.org>

YevheniiHladkevych committed:
Merge remote-tracking branch 'origin/feature/parse-given-step-on-feature-level' into feature/parse-given-step-on-feature-level

# Conflicts:
#	src/main/kotlin/com/networkedassets/gherkin/runner/runners/CallbackRunner.kt
#	src/main/kotlin/com/networkedassets/gherkin/runner/runners/FeatureRunner.kt
#	src/main/kotlin/com/networkedassets/gherkin/runner/specification/FeatureSpecification.kt
#	src/main/kotlin/com/networkedassets/gherkin/runner/util/GherkinConverter.kt

YevheniiHladkevych committed:
Fix the build

Signed-off-by: YevheniiHladkevych <yhladkevych@networkedassets.org>

YevheniiHladkevych committed:
Commit file with conflicts

YevheniiHladkevych committed:
Test jitpack errors

YevheniiHladkevych committed:
Debug #1

YevheniiHladkevych committed:
Works when all scenarios contain Background

YevheniiHladkevych committed:
Works for empty background use case

YevheniiHladkevych committed:
Add null safety

YevheniiHladkevych committed:
Uncomment JUnit exporter

YevheniiHladkevych committed:
Remove unused function

YevheniiHladkevych committed:
Small refactoring

YevheniiHladkevych committed:
Made GherkinRunner store List of backgrounds instead of pure data

YevheniiHladkevych committed:
Use mutableListOf<T> instead of List<T>

tlazarski commented on src/main/kotlin/com/networkedassets/gherkin/runner/runners/CallbackRunner.kt:
```diff
@@ -32,7 +32,7 @@ class CallbackRunner(private val callbackDescription: Description,
         try {
             val callbackMethod = Reflection.getCallbackMethod(featureSpecification, callbackType)
-            runImplemented(featureSpecification, callbackMethod)
+             runImplemented(featureSpecification, callbackMethod)
```
I think this space is not needed

YevheniiHladkevych committed:
Use wrapper GherkinBackground object

YevheniiHladkevych committed:
Remove unnecessary method

YevheniiHladkevych committed:
Add utility methods

YevheniiHladkevych committed:
Add private modifier

YevheniiHladkevych committed:
Fix Background with empty title issue

- if background name is empty, ignores it in reports
- fix Examples starting from hash sign
- add test cases

YevheniiHladkevych committed:
SBP-TAF-43

- allows Hash sign in examples with additional type references

YevheniiHladkevych committed:
SBP-TAF-25: Implement basic Environmental Bindings loading

- bindings from env data table are available under "envBinding" property
- If first row in Env datatable contains hash signs -> ignore first row

YevheniiHladkevych committed:
SBP-TAF-25: Fix bug with cropping feature steps

YevheniiHladkevych committed:
SBP-TAF-25: Fix bug tenary if

tlazarski commented on src/main/kotlin/com/networkedassets/gherkin/runner/util/GherkinConverter.kt:
```diff
@@ -64,6 +63,19 @@ object GherkinConverter {
+        val converted = mutableMapOf<String, String>()
+        dataTableForStep
+                ?.dropWhile { row -> row[0].contains("#") }
+                ?.map { row -> converted.put(row[0], row[1]) }
```
Using map for mutating object, which is out of scope of mapping function is bad idea, which is misleading. You should use foreach in this case. There is also other approach. You can use associate() method to map list to Map and then assign it to "converted" variable:
.associate{ row  -> row[0] to row[1] }

tlazarski commented on src/main/kotlin/com/networkedassets/gherkin/runner/util/GherkinConverter.kt:
```diff
@@ -64,6 +63,19 @@ object GherkinConverter {
+    }
+
+    private fun convertBackground(feature: Feature): GherkinBackground {
+        val filteredBackgr = feature.children.filterIsInstance<Background>()
```
I think, that simple name "backgrounds" will be better than "filteredBackgr"

tlazarski commented on src/main/kotlin/com/networkedassets/gherkin/runner/util/GherkinConverter.kt:
```diff
@@ -1,34 +1,31 @@
 object GherkinConverter {
     fun convertFeature(feature: Feature): GherkinFeature {
-        val gherkinFeature = GherkinFeature(feature.name, feature.tags.map { it.name })
+        val envBindings = ExampleBindings(convert2DArrToEnvBindings(convertBackground(feature).getDataTableForStep(0)))
```
convertBackground(feature) used twice in the same scope

tlazarski commented on src/main/kotlin/com/networkedassets/gherkin/runner/util/GherkinConverter.kt:
```diff
@@ -1,34 +1,31 @@
+        val envBindings = ExampleBindings(convert2DArrToEnvBindings(convertBackground(feature).getDataTableForStep(0)))
+        val gherkinFeature = GherkinFeature(feature.name,
+                feature.tags.map { it.name },
+                convertBackground(feature),
```
convertBackground(feature) used twice in the same scope

YevheniiHladkevych committed:
SBP-TAF-25: Fix according the suggestions

- check for Background keyword on the conversion step not on the runner step
- fix map associate
